File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        lo,hi=0,len(nums)-1
        res=[-1,-1]
        while lo<=hi:
            mid=lo+((hi-lo)>>1);
            if nums[mid]==target:
                res[0]=mid
                hi=mid-1
            elif nums[mid]>target:
                hi=mid-1
            else:
                lo=mid+1
        lo,hi=0,len(nums)-1
        while lo<=hi:
            mid=lo+((hi-lo)>>1);
            if nums[mid]==target:
                res[1]=mid
                lo=mid+1
            elif nums[mid]>target:
                hi=mid-1
            else:
                lo=mid+1
        return res
        

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.cpp

/* 
    Time: O(log(n))
    Space: O(1)
    Tag: Binary Search
    Difficulty: M
*/

#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

class Solution {
    int bSearchMin(vector<int> nums, int target) {
        int mid, l = 0, h = nums.size() - 1;
        while (l <= h) {
            mid = (h - l) / 2 + l;
            if (nums[mid] == target && (mid == 0 || nums[mid - 1] != target)) {
                return mid;
            } else if (nums[mid] < target) {
                l = mid + 1;
            } else {
                h = mid - 1;
            }
        }
        return -1;
    }
    int bSearchMax(vector<int> nums, int target) {
        int mid, l = 0, h = nums.size() - 1;
        while (l <= h) {
            mid = (h - l) / 2 + l;
            if (nums[mid] == target && ((mid == nums.size() - 1) || nums[mid + 1] != target)) {
                return mid;
            } else if (nums[mid] > target) {
                h = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return -1;
    }

public:
    vector<int> searchRange(vector<int> &nums, int target) {
        return {bSearchMin(nums, target), bSearchMax(nums, target)};
    }
};